modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,595 @@
+
+private:
+    void updateWeights(InputArray sample, bool isFirstClass, float gamma, Mat &weights);
+
+    std::pair<bool,bool> areClassesEmpty(Mat responses);
+
+    void writeParams( FileStorage &fs ) const;
+
+    void readParams( const FileNode &fn );
+
+    static inline bool isFirstClass(float val) { return val > 0; }
+
+    static void normalizeSamples(Mat &matrix, Mat &average, float &multiplier);
+
+    float calcShift(InputArray _samples, InputArray _responses) const;
+
+    static void makeExtendedTrainSamples(const Mat &trainSamples, Mat &extendedTrainSamples, Mat &average, float &multiplier);
+
+
+
+    // Vector with SVM weights
+    Mat weights_;
+    float shift_;
+
+    // Parameters for learning
+    struct SVMSGDParams
+    {
+        float lambda;                             //regularization
+        float gamma0;                             //learning rate
+        float c;
+        TermCriteria termCrit;
+        SvmsgdType svmsgdType;
+        MarginType marginType;
+    };
+
+    SVMSGDParams params;
+};
+
+Ptr<SVMSGD> SVMSGD::create()
+{
+    return makePtr<SVMSGDImpl>();
+}
+
+std::pair<bool,bool> SVMSGDImpl::areClassesEmpty(Mat responses)
+{
+    CV_Assert(responses.cols == 1 || responses.rows == 1);
+    std::pair<bool,bool> emptyInClasses(true, true);
+    int limit_index = responses.rows;
+
+    for(int index = 0; index < limit_index; index++)
+    {
+        if (isFirstClass(responses.at<float>(index)))
+            emptyInClasses.first = false;
+        else
+            emptyInClasses.second = false;
+
+        if (!emptyInClasses.first && ! emptyInClasses.second)
+            break;
+    }
+
+    return emptyInClasses;
+}
+
+void SVMSGDImpl::normalizeSamples(Mat &samples, Mat &average, float &multiplier)
+{
+    int featuresCount = samples.cols;
+    int samplesCount = samples.rows;
+
+    average = Mat(1, featuresCount, samples.type());
+    for (int featureIndex = 0; featureIndex < featuresCount; featureIndex++)
+    {
+        Scalar scalAverage = mean(samples.col(featureIndex))[0];
+        average.at<float>(featureIndex) = static_cast<float>(scalAverage[0]);
+    }
+
+    for (int sampleIndex = 0; sampleIndex < samplesCount; sampleIndex++)
+    {
+        samples.row(sampleIndex) -= average;
+    }
+
+    double normValue = norm(samples);
+
+    multiplier = static_cast<float>(sqrt(samples.total()) / normValue);
+
+    samples *= multiplier;
+}
+
+void SVMSGDImpl::makeExtendedTrainSamples(const Mat &trainSamples, Mat &extendedTrainSamples, Mat &average, float &multiplier)
+{
+    Mat normalisedTrainSamples = trainSamples.clone();
+    int samplesCount = normalisedTrainSamples.rows;
+
+    normalizeSamples(normalisedTrainSamples, average, multiplier);
+
+    Mat onesCol = Mat::ones(samplesCount, 1, CV_32F);
+    cv::hconcat(normalisedTrainSamples, onesCol, extendedTrainSamples);
+}
+
+void SVMSGDImpl::updateWeights(InputArray _sample, bool firstClass, float gamma, Mat& weights)
```

Comment:
Only at this point I figured that firstClass and secondClass are the classes of the classifier. Why not to use a more conventional naming - positives and negatives? Or maybe there's another conventional name.